lsp.go

```diff
@@ -230,59 +225,72 @@ func (s *Server) handleDocumentSaved(
 	}
 
 	fname := params.TextDocument.URI.Filename()
-	data, err := os.ReadFile(fname)
+	content, err := os.ReadFile(fname)
 	if err != nil {
 		log.Error().Err(err).Msg("reading saved file.")
 		return nil
 	}
 
-	err = checkFile(fname, string(data))
-	return reply(ctx, nil,
-		s.sendErrorDiagnostics(ctx, params.TextDocument.URI, err),
-	)
+	return s.checkAndReply(ctx, reply, fname, string(content))
 }
 
-func (s *Server) sendErrorDiagnostics(ctx context.Context, currentFile lsp.URI, err error) error {
-	if err == nil {
-		log.Debug().Str("file", currentFile.Filename()).
-			Msg("cleaning editor errors")
-
-		// this is required to clear the `problems panel` for the active file
-		// if it had errors before.
-		s.sendDiagnostics(ctx, currentFile, []lsp.Diagnostic{})
-		return nil
+func (s *Server) sendErrorDiagnostics(ctx context.Context, files []string, err error) error {
+	errs := errors.L()
+	switch e := err.(type) {
+	case *errors.Error:
+		// an error can wrap a list of errors.
+		errs = e.AsList()
+	case *errors.List:
+		errs = e
+	default:
+		if err != nil {
+			log.Debug().Err(err).Msg("unknown error ignored because it doesn't provide file range")
+		}
```

Comment:
we cannot bail out if error is nil because we have to report that in this case all files in the directory are ok. In order to do that we need to send an empty list of diagnostics.

That's why we have the `diagsMap`. All the time this function is called, it should send a diagnostic for all the files, the ones not in the errors.List gets an empty slice of diagnostic.